### PR TITLE
[MSHARED-884] - Don't always overwrite filtered resources

### PR DESCRIPTION
**Pull request by roxspring**
- Added unit tests for `FileUtils.copyFile()` methods using `FilterWrapper`
- When filtering and _potentially_ overwriting an existing file:
  - Open the existing file as a `RandomAccessFile`
  - Encode each buffer of content
  - Compare encoded buffer with a matching buffer from the existing file
  - If different then overwrite the remainder of the file  
- Added unit tests to demonstrate that lastModified date is only updated when the content changes.
- Rudimentary performance tests with various source sizes show no significant impact:

  Source Size | Avg Millis (Old) | Avg Millis (New)
  -- | -- | --
  100 MB | 6034 | 5940
  1 MB | 54 | 56
  10 KB | 1 | 8

Relates to MSHARED-884

### CODE REVIEW

File: src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java [hunk 1]
```diff
@@ -1,29 +1,49 @@
...
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.junit.Assume.assumeFalse;
+import static org.junit.Assume.assumeThat;
+import static org.junit.Assume.assumeTrue;
```
Comment (roxspring) [inline src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java:46]: Looks like Intellij moved some imports when I wasn't looking. Happy to fix if it's an issue.

File: src/main/java/org/apache/maven/shared/utils/io/FileUtils.java [hunk 2]
```diff
@@ -1826,34 +1831,101 @@ public static void copyFile( @Nonnull File from, @Nonnull File to, @Nullable Str
                                  @Nullable FilterWrapper[] wrappers, boolean overwrite )
         throws IOException
     {
-        if ( wrappers != null && wrappers.length > 0 )
+        if ( wrappers == null || wrappers.length <= 0 )
```
Comment (roxspring) [inline src/main/java/org/apache/maven/shared/utils/io/FileUtils.java:1834]: Flipped the condition around to get the simple case out of the way and ease the cognitive load when reading the else clause. Maybe I should split this out into a separate commit and make that final check `==` since negative lengths shouldn't be possible!

File: src/main/java/org/apache/maven/shared/utils/io/FileUtils.java [hunk 3]
```diff
@@ -109,7 +114,7 @@ protected FileUtils()
     /**
      * The file copy buffer size (30 MB)
      */
-    private static final long FILE_COPY_BUFFER_SIZE = ONE_MB * 30;
+    private static final int FILE_COPY_BUFFER_SIZE = ONE_MB * 30;
```
Comment (roxspring) [inline src/main/java/org/apache/maven/shared/utils/io/FileUtils.java:117]: Don't think there was any need for this to be a `long` other than the only previous use case, where it's now automatically cast to a long anyway.

File: src/main/java/org/apache/maven/shared/utils/io/FileUtils.java [hunk 4]
```diff
@@ -1826,34 +1831,101 @@ public static void copyFile( @Nonnull File from, @Nonnull File to, @Nullable Str
...
+                    }
+                }
+                else
+                {
+                    CharsetEncoder encoder = Charset.forName( encoding ).newEncoder();
+
+                    int totalBufferSize = FILE_COPY_BUFFER_SIZE;
+
+                    int charBufferSize = ( int ) Math.floor( totalBufferSize / ( 2 + 2 * encoder.maxBytesPerChar() ) );
+                    int byteBufferSize = ( int ) Math.ceil( charBufferSize * encoder.maxBytesPerChar() );
```
Comment (roxspring) [inline src/main/java/org/apache/maven/shared/utils/io/FileUtils.java:1872]: This is attempting to consume the same amount of buffer memory in total, but split between a `char` buffer and a pair of `byte` buffers - each big enough to contain the encoded `char` buffer.

File: src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java [hunk 5]
```diff
@@ -438,9 +448,300 @@ public void copyFile1()
             testFile1.lastModified() == destination.lastModified());*/
     }
 
+    private static long MILLIS_PER_DAY = 24 * 60 * 60 * 1000L;
+
+    /** A time today, rounded down to the previous minute */
+    private static long MODIFIED_TODAY = (System.currentTimeMillis() / 60_000) * 60_000;
```
Comment (roxspring) [inline src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java:454]: Throwing away seconds here allows these precise times to be used accurately even on low resolution file systems such as FAT32.

**Comment by michael-o** [comment]
What is the general benefit you expect here? Not changing last modified date?

**Comment by roxspring** [comment]
Yes. That’s the point, as discussed in the jira issue and dev mailing lists. Unnecessarily  modifying the file when the filter content hasn’t changed can lead to wasted work downstream.

**Comment by michael-o** [comment]
Please change the buffer code to: https://github.com/apache/maven-wagon/commit/92c0d2a858a38d9b3d0aefc0f9298d5c31c7e508

**Comment by roxspring** [comment]
@michael-o I think I've fixed the Buffer code as you requested - I'd be grateful of another review!

File: src/main/java/org/apache/maven/shared/utils/io/FileUtils.java [hunk 6]
```diff
@@ -1810,34 +1816,101 @@ public static void copyFile( @Nonnull File from, @Nonnull File to, @Nullable Str
...
-            }
-        }
-        else
-        {
-            if ( to.lastModified() < from.lastModified() || overwrite )
-            {
-                copyFile( from, to );
+                if ( overwrite || !to.exists() )
+                {
+                    try ( Writer fileWriter = new OutputStreamWriter( new FileOutputStream( to ), encoding ) )
```
Comment (olamy) [inline src/main/java/org/apache/maven/shared/utils/io/FileUtils.java:1845]: what about `Files.newBufferedWriter(  )`

File: src/main/java/org/apache/maven/shared/utils/io/FileUtils.java [hunk 7]
```diff
@@ -1810,34 +1816,101 @@ public static void copyFile( @Nonnull File from, @Nonnull File to, @Nullable Str
...
             {
                 encoding = Charset.defaultCharset().name();
             }
-            
+
             // buffer so it isn't reading a byte at a time!
             try ( Reader fileReader =
-                new BufferedReader( new InputStreamReader( new FileInputStream( from ), encoding ) );
-                            Writer fileWriter = new OutputStreamWriter( new FileOutputStream( to ), encoding ) )
+                    new BufferedReader( new InputStreamReader( new FileInputStream( from ), encoding ) ) )
```
Comment (olamy) [inline src/main/java/org/apache/maven/shared/utils/io/FileUtils.java:1835]: what about `Files.newBufferedReader(  )`?

Comment (roxspring) [reply src/main/java/org/apache/maven/shared/utils/io/FileUtils.java]: Done.

File: src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java [hunk 8]
```diff
@@ -441,9 +449,300 @@ public void copyFile1()
             testFile1.lastModified() == destination.lastModified());*/
     }
 
+    private static long MILLIS_PER_DAY = 24 * 60 * 60 * 1000L;
```
Comment (michael-o) [inline src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java:452]: Why not use TimeUnit for this?

Comment (roxspring) [reply src/test/java/org/apache/maven/shared/utils/io/FileUtilsTest.java:452]: Happy to use `TimeUnit.DAYS.toMillis(1)` instead. Assumed `ChronoUnit` and friends weren't available but had forgotten `TimeUnit`!

**Comment by roxspring** [comment]
> Does it actually make sense to introduce a threshold? E.g., > 1 MB?

You're suggesting that we apply the new logic only to input >1MB and for smaller files we presumably process them in memory and compare that to the existing written file, right?

In general I dislike maintaining multiple equivalent codepaths and switching between them conditionally as typically one is used far more often and the others languish in bit rot, becoming a maintenance burden. Sometimes this approach is necessary to gain performance in different circumstances, but is that clearly the case here?

I'm not sure it is. For a start I don't remember what rounding errors might have been involved in the 1 vs 8 millisecond comparison from the PR description, but more importantly that's comparing broken to fixed behaviour - the 1 millisecond figure timing is for the existing "unconditionally overwrite when filtering", and we don't yet have comparable numbers for "process them in memory and compare that to the existing written file".

If an additional 7 milliseconds per filtered resource is enough to cause panic then I guess I can look into a thresholded approach but my inclination is to avoid that effort until it proves necessary.

**Comment by michael-o** [comment]
> 
> 
> > Does it actually make sense to introduce a threshold? E.g., > 1 MB?
> 
> You're suggesting that we apply the new logic only to input >1MB and for smaller files we presumably process them in memory and compare that to the existing written file, right?
> 
> In general I dislike maintaining multiple equivalent codepaths and switching between them conditionally as typically one is used far more often and the others languish in bit rot, becoming a maintenance burden. Sometimes this approach is necessary to gain performance in different circumstances, but is that clearly the case here?
> 
> I'm not sure it is. For a start I don't remember what rounding errors might have been involved in the 1 vs 8 millisecond comparison from the PR description, but more importantly that's comparing broken to fixed behaviour - the 1 millisecond figure timing is for the existing "unconditionally overwrite when filtering", and we don't yet have comparable numbers for "process them in memory and compare that to the existing written file".
> 
> If an additional 7 milliseconds per filtered resource is enough to cause panic then I guess I can look into a thresholded approach but my inclination is to avoid that effort until it proves necessary.

Well reasoned, let's leave it as-is.

**Comment by olamy** [comment]
merged via e57180309a64758e161513d80f1185e13dc84826
